graphula-core/src/Graphula.hs

```diff
@@ -156,14 +204,24 @@ runGraphulaReplay
   -> Graph FromJSON NoConstraint nodeConstraint entity m a -> m a
 runGraphulaReplay replayFile frontend f = do
   replayLog <- liftIO $ newIORef =<< (lines <$> readFile replayFile)
-  catch (go replayLog) (rethrowHUnitReplay replayFile)
-  where
-    go replayLog =
-      flip iterT f $ \case
-        InR r -> frontend r
-        InL l -> backendReplay replayLog l
-
-
+  runGraphulaUsing (backendReplay replayLog) frontend f
+    `catch` rethrowHUnitReplay replayFile
+
+
+finalizerFrontend
+  :: (MonadThrow m, MonadIO m)
+  => IORef (FreeT (Frontend nodeConstraint entity) m ())
+  -> Frontend nodeConstraint entity (m a)
+  -> FreeT (Frontend nodeConstraint entity) m a
+finalizerFrontend finalizersRef f = case f of
+  Insert n next -> do
+    mEnt <- liftF $ Insert n id
+    for_ mEnt $ \ent ->
+      liftIO $ modifyIORef' finalizersRef ((>>) (liftF $ Remove ent ()))
```

Comment:
Yeah, that is more readable. I took it a step further and abstracted the `liftF`.